### PR TITLE
Performance improvement: Change Ray dispatch

### PR DESCRIPTION
**Pull request by PMunkes**
This change improves rendering performance on an RX 6800 by 3-5%.
It changes the raytracing dispatches to only collect a single sample per ray generation shader invocation. This reduces the execution time of every wavefront and increases the amount of threads. This is done by dispatching threads in the z-direction.

| Dispatch | Samples | Scene 1 | Scene 2 | Scene 3 | Scene 4 | Scene 5 |
|----------|---------|---------|---------|---------|---------|---------|
| new      | 2       | 145.41  | 143.44  | 68.91   | 113.5   | 42.5    |
| old      | 2       | 140.04  | 138.45  | 66      | 108.66  | 40.16   |
| new      | 4       | 74.72   | 73.85   | 35.08   | 58.05   | 21.49   |
| old      | 4       | 72.63   | 71.73   | 33.98   | 55.86   | 20.43   |
| new      | 8       | 38.01   | 37.48   | 17.71   | 29.35   | 10.79   |
| old      | 8       | 36.83   | 36.49   | 17.27   | 28.32   | 10.33   |
| new      | 16      | 19.19   | 18.9    | 8.9     | 14.79   | 5.42    |
| old      | 16      | 18.66   | 18.47   | 8.73    | 14.29   | 5.22    |

Performance improvements with new dispatch compared to old.

| Samples | Scene 1 | Scene 2 | Scene 3 | Scene 4 | Scene 5 |
|---------|---------|---------|---------|---------|---------|
| 2       | 103.83% | 103.60% | 104.41% | 104.45% | 105.83% |
| 4       | 102.88% | 102.96% | 103.24% | 103.92% | 105.19% |
| 8       | 103.19% | 102.71% | 102.57% | 103.63% | 104.50% |
| 16      | 102.84% | 102.33% | 101.95% | 103.50% | 103.83% |

### CODE REVIEW

**Comment by PMunkes** [comment]
This is mostly inspired by my investigations in issue #28. 

I was tinkering around with some supersampling code for the last few weeks, but I never could get it to work just right.

**Comment by PMunkes** [comment]
I'm thinking of revisiting the original idea in a few weeks. I think using `gl_LaunchSizeEXT.z ` would avoid most of the issues I had with the original supersampling version of the patch.

File: assets/shaders/RayTracing.rgen [hunk 1]
```diff
@@ -25,79 +25,74 @@ void main()
...
+
+		// Trace hit.
+		origin = origin + t * direction;
+		direction = vec4(Ray.ScatterDirection.xyz, 0);
+	} 
 
-	const bool accumulate = Camera.NumberOfSamples != Camera.TotalNumberOfSamples;
-	const vec3 accumulatedColor = (accumulate ? imageLoad(AccumulationImage, ivec2(gl_LaunchIDEXT.xy)) : vec4(0)).rgb + pixelColor;
+	//Only reset the image on the first thread touching it.
+	const bool accumulate = Camera.NumberOfSamples != Camera.TotalNumberOfSamples
```
Comment (GPSnoopy) [inline assets/shaders/RayTracing.rgen:78]: If it's anything like in CUDA, this is likely a race condition as you have no guarantee that the `z == 0` threads are run before the `z != 0` threads. So you are probably losing samples on the first image (when resetting the accumulation buffer).

Unless the `vkCmdTraceRaysKHR` called in a loop with the incremented depth is accounting for that? Not sure if it requires any memory barrier then.

Comment (PMunkes) [reply assets/shaders/RayTracing.rgen:78]: Yes, this is indeed a race condition, but the only consequence is potentially loosing a few random samples on the debug view and on the first frame after reset.

There might be a way to do this properly by introducing an VK_FORMAT_R32_UINT texture to collect the samples per pixel, but I'm unsure if it's worth it if this race condition isn't triggered in practice.

Comment (PMunkes) [reply assets/shaders/RayTracing.rgen:78]: I just remembered, that I still should have code for that around, I will try to implement it.

**Comment by PMunkes** [comment]
I've now done a lot more research. As it turns out, there is no strict ordering of imageLoads and stores, so this actually never worked properly. Only the last store actually reached memory, so at best this always only displayed the last sample.

**Comment by PMunkes** [comment]
Final note: There could be a way to actually do this, by first rendering 1 sample per pixel onto a large texture and then rendering that out onto a full screen triangle. This would produce an effect similar to what I tried to archieve here, but it would complicate things like changing the sample count ore frame.

**Pull request closed without merging**